Support signing with capabilities

tomsmalley opened the pull request:
Adds support for capabilities. The signing API passes all capabilities
it needs, and these are displayed to the user with an account dropdown
box. Other deployment methods allow the user to specify arbitrary
capabilities.

The sender/gas payer field has not been altered here, but in future
could be moved to the capabilities/roles section too.

This also fixes the transaction hash stuff in deployment dialogs. We
were hashing the code, not the command.

### **Don't merge this yet, this isn't a guaranteed change upstream**

tomsmalley committed:
Support signing with capabilities

Adds support for capabilities. The signing API passes all capabilities
it needs, and these are displayed to the user with an account dropdown
box. Other deployment methods allow the user to specify arbitrary
capabilities.

The sender/gas payer field has not been altered here, but in future
could be moved to the capabilities/roles section too.

This also fixes the transaction hash stuff in deployment dialogs. We
were hashing the code, not the command.

tomsmalley committed:
Use newer signing-api

mightybyte committed:
Add missing dep

tomsmalley committed:
Use signing API proxy as defined upstream

alexfmpe commented on frontend/src/Frontend/UI/DeploymentSettings.hs:
```diff
@@ -572,21 +579,87 @@ uiChainSelection info cls = mdo
+        False -> do
+          ma <- account
+          e <- parsed
+          pure $ ffor ma $ \a -> Map.singleton a (either (const []) pure e)
```
I think `pure $ ffor ma $ \a -> Map.singleton a (toList e)` does the same - unsure of which is clearer.

tomsmalley replied:
I feel like `either` is clearer, but this is totally subjective.

alexfmpe commented on frontend/src/Frontend/UI/DeploymentSettings.hs:
```diff
@@ -572,21 +579,87 @@ uiChainSelection info cls = mdo
+      & modifyAttributes .~ ffor errors (\e -> "style" =: ("background-color: #fdd" <$ guard e))
+    empty <- holdUniqDyn $ T.null <$> value cap
+    let parsed = parseSigCapability <$> value cap
+        hasError = (\p e -> isLeft p && not e) <$> parsed <*> empty
```
Does `empty` imply `parseSigCapability` return `Left` ?

tomsmalley replied:
yeah, this could be outdated now. I changed the implementation midway and this might be a leftover

tomsmalley replied:
Actually, I think this is right. Yes, `parseSigCapability` returns `Left` when given no input, but we don't want the error to display for empty inputs.

alexfmpe replied:
Maybe rename to `showError` ?

alexfmpe commented on frontend/src/Frontend/UI/DeploymentSettings.hs:
```diff
@@ -572,21 +579,87 @@ uiChainSelection info cls = mdo
+  pure
+    ( empty
+    , do
+      empty >>= \case
```
```haskell
ffor3 empty account parsed $ \case
  True -> Just mempty
  False -> \ma e -> ffor ma $ \a -> Map.singleton a (either (const []) pure e)
```

tomsmalley replied:
Resolving based on conversation on mattermost

alexfmpe commented on frontend/src/Frontend/UI/DeploymentSettings.hs:
```diff
@@ -572,21 +579,87 @@ uiChainSelection info cls = mdo
+        elClass "th" "table__heading" $ text "Role"
+        elClass "th" "table__heading" $ text "Capability"
+        elClass "th" "table__heading" $ text "Account"
+      let fixup = fmap (fmap (Map.unionsWith (<>)) . sequence) . sequence
```
What does the `Maybe` in `Maybe AccountName` mean?

IIUC, the inner `sequence` is 
```
   [Maybe (Map AccountName [SigCapability])] 
-> Maybe [(Map AccountName [SigCapability])]
```
so any `Nothing` in the list will cause the whole thing to be `Nothing`

Why do we want that rather than dropping the failures, e.g. 
```
fmapMaybe id 
:: [Maybe (Map AccountName [SigCapability])] 
-> [(Map AccountName [SigCapability])]
```

tomsmalley replied:
I didn't want to allow submission of a TX by silently dropping partial capabilities

tomsmalley committed:
WIP: Bump servant...

Commit:
Allow for selection of different user accounts for capabilities

tomsmalley committed:
Fix public key column in deployment preview

tomsmalley committed:
Make signing API dialog send responses when cancelled

tomsmalley committed:
Fix warnings

alexfmpe committed:
Fix 'frontend' frontend build

alexfmpe committed:
Fix 'desktop' build

Commit:
Fix up account select for capabilities and remove separation for gas payer and add it as a row to the roles table

Commit:
Change capability for gas to FUND_TX

tomsmalley committed:
Rename hasError -> showError

tomsmalley committed:
Merge pull request #152 from kadena-io/sc-user-select-for-caps

user select for capabilities

tomsmalley committed:
Merge remote-tracking branch 'origin/ts-signing-api-fixes' into sig-with-caps

tomsmalley committed:
Merge pull request #153 from kadena-io/ts-signing-api-fixes

Signing api fixes

alexfmpe committed:
Remove ghcid dependencies from project shells

tomsmalley committed:
Merge branch 'sig-with-caps' of github.com:kadena-io/chainweaver into sig-with-caps

tomsmalley committed:
Bump signing-api, support extra-senders

tomsmalley committed:
Make extraSigners optional

Commit:
Return validation information as well as values for caps widgets.

The `capabilityInputRow(s)` functions now return data types that carry information
regarding empty and error states. This lets the parent widget decide what to do in the
case of errors or empty fields.

Not convinced this is the best way to solve this.

Commit:
Remove debug import and comment, add some useful comments

tomsmalley committed:
Merge pull request #160 from kadena-io/ts-extra-senders

Add extraSigners

alexfmpe committed:
Bump pact

alexfmpe committed:
Merge branch 'sig-with-caps' into ts-servant-bump

alexfmpe committed:
Add missing import

tomsmalley committed:
Use result of parsed cap to determine errors

tomsmalley committed:
No more debugging

tomsmalley committed:
Merge pull request #163 from kadena-io/ts-more-caps

Capability fun

tomsmalley committed:
Merge branch 'sig-with-caps' into ts-servant-bump

tomsmalley committed:
Fix warnings

tomsmalley committed:
Merge pull request #161 from kadena-io/ts-servant-bump

Pact bump
